src/profile-logic/call-tree.js

```diff
@@ -523,22 +630,155 @@ export function computeCallNodeLeafAndSummary(
   return { callNodeLeaf, rootTotalSummary };
 }
 
+export function getSelfAndTotalForCallNode(
+  callNodeIndex: IndexIntoCallNodeTable,
+  callNodeInfo: CallNodeInfo,
+  callTreeTimings: CallTreeTimings
+): SelfAndTotal {
+  switch (callTreeTimings.type) {
+    case 'NON_INVERTED': {
+      const { timings } = callTreeTimings;
+      const self = timings.self[callNodeIndex];
+      const total = timings.total[callNodeIndex];
+      return { self, total };
+    }
+    case 'INVERTED': {
+      const callNodeInfoInverted = ensureExists(callNodeInfo.asInverted());
+      const { timings } = callTreeTimings;
+      const { callNodeLeaf, totalPerRootNode } = timings;
+      if (callNodeInfoInverted.isRoot(callNodeIndex)) {
+        const total = totalPerRootNode.get(callNodeIndex) ?? 0;
+        return { self: total, total };
+      }
+      const { total } = _getInvertedTreeNodeTotalAndHasChildren(
+        callNodeIndex,
+        callNodeInfoInverted,
+        callNodeLeaf
+      );
+      return { self: 0, total };
+    }
+    default:
+      throw assertExhaustiveCheck(callTreeTimings.type);
+  }
+}
+
+function _getInvertedTreeNodeTotalAndHasChildren(
+  callNodeIndex: IndexIntoCallNodeTable,
+  callNodeInfo: CallNodeInfoInverted,
+  callNodeLeaf: Float32Array
+): TotalAndHasChildren {
+  const nodeDepth = callNodeInfo.getCallNodeTable().depth[callNodeIndex];
+  const [rangeStart, rangeEnd] =
+    callNodeInfo.getSuffixOrderIndexRangeForCallNode(callNodeIndex);
+  const suffixOrderedCallNodes = callNodeInfo.getSuffixOrderedCallNodes();
+  const callNodeTableDepthCol =
+    callNodeInfo.getNonInvertedCallNodeTable().depth;
+  let total = 0;
+  let hasChildren = false;
+  for (let i = rangeStart; i < rangeEnd; i++) {
+    const leafNode = suffixOrderedCallNodes[i];
+    const leaf = callNodeLeaf[leafNode];
+    total += leaf;
+
+    // The inverted call node has children if any deeper call paths with non-zero
+    // self time contribute to it.
+    hasChildren =
+      hasChildren ||
+      (leaf !== 0 && callNodeTableDepthCol[leafNode] > nodeDepth);
```

Comment:
This is definitely confusing, because we are dealing with both inverted and non-inverted call nodes here.

`callNodeIndex` is a node in the **inverted** tree.
`leafNode` is a node in the **non-inverted** tree.
`callNodeTableDepthCol` is the depth column of the **non-inverted** call node table.

It's easiest to look at the example from the `CallNodeInfoInverted` docs:


## Example

### Non-inverted call tree:

Legend:

cnX: Non-inverted call node index X
soX: Suffix order index X

```
  Tree            Left aligned    Right aligned        Reordered by suffix
- [cn0] A      =  A            =            A [so0]    [so0] [cn0] A
  - [cn1] B    =  A -> B       =       A -> B [so3]    [so1] [cn4] A <- A
    - [cn2] A  =  A -> B -> A  =  A -> B -> A [so2] ↘↗ [so2] [cn2] A <- B <- A
    - [cn3] C  =  A -> B -> C  =  A -> B -> C [so6] ↗↘ [so3] [cn1] B <- A
  - [cn4] A    =  A -> A       =       A -> A [so1]    [so4] [cn5] B <- A <- A
    - [cn5] B  =  A -> A -> B  =  A -> A -> B [so4]    [so5] [cn6] C <- A
  - [cn6] C    =  A -> C       =       A -> C [so5]    [so6] [cn3] C <- B <- A
```

### Inverted call tree:

Legend, continued:

inX:     Inverted call node index X
so:X..Y: Suffix order index range soX..soY (soY excluded)

```
                                                Represents call paths ending in
- [in0] A  (so:0..3)        =  A             =            ... A (cn0, cn4, cn2)
  - [in1] A  (so:1..2)      =  A <- A        =       ... A -> A (cn4)
  - [in2] B  (so:2..3)      =  A <- B        =       ... B -> A (cn2)
    - [in3] A  (so:2..3)    =  A <- B <- A   =  ... A -> B -> A (cn2)
- [in4] B  (so:3..5)        =  B             =            ... B (cn1, cn5)
  - [in5] A  (so:3..5)      =  B <- A        =       ... A -> B (cn1, cn5)
    - [in6] A  (so:4..5)    =  B <- A <- A   =  ... A -> A -> B (cn5)
- [in7] C  (so:5..7)        =  C             =            ... C (cn6, cn3)
  - [in8] A  (so:5..6)      =  C <- A        =       ... A -> C (cn6)
  - [in9] B  (so:6..7)      =  C <- B        =       ... B -> C (cn3)
    - [in10] A  (so:6..7)   =  C <- B <- A   =  ... A -> B -> C (cn3)
```

Let's say `callNodeIndex` is `in5`. The total time for `in5` is the self time of `cn1` plus the self time of `cn5`.
And `in5` has children, because `cn5` has a depth of 3, which is more than the depth of `in5`.

Conversely, `in1` has no children - the only non-inverted call node which contributes to it, `cn4`, has depth 2.

I'll respond to your questions below anyway.

> I don't understand the second part. Aren't all the call nodes with selftime between rangeStart and rangeEnd children?

You can't really say "they are children" because those nodes are non-inverted nodes, and "children" here refers to inverted nodes.

> Actually I think I don't understand the first part either: why are the children in an inverted tree the ones with self time?

The children in the inverted tree are inverted nodes with non-zero total time, and we compute the total time of an inverted node by summing up the self time of the non-inverted nodes that contribute to that inverted node.

---

I'm changing the comment to the following:

```js
    // The inverted call node has children if it has any inverted child nodes
    // with non-zero total time. The total time of such an inverted child node
    // is the sum of the self times of the non-inverted call nodes which
    // contribute to it. Does `leafNode` contribute to one of our children?
    // Maybe. To do so, it would need to describe a call path whose length is at
    // least as long as the inverted call paths of our children - if not, it only
    // contributes to `callNodeIndex` and not to our children.
    // Rather than comparing the length of the call paths, we can just compare
    // the depths.
    //
    // In other words:
    // The inverted call node has children if any deeper call paths with non-zero
    // self time contribute to it.
```